File: proposed-projects/round-trip-experiment/reverse/20260909T132655Z_2a16d72b/20260910T075017Z_244b957c/ported_implementation/src/rules-builder/symbol-ids.ts
export class SymbolIds {
  // we don't need to delete, just preserve relationships
  private _map = new Map<string, number>();
  private _pos = new Map<string, number>();
  private _reverseMap = new Map<number, string>();

  get size(): number {
    return this._map.size;
  }

  keys(): IterableIterator<string> {
    return this._map.keys();
  }

  has(key: string): boolean {
    return this._map.has(key);
  }

  get(key: string): number {
    if (!this._map.has(key)) {
      throw new Error(`SymbolIds does not contain key: ${key}`);
    }
    return this._map.get(key) as number;
  }

  getOrUndefined(key: string): number | undefined {
    return this._map.get(key);
  }

  reverseGet(key: number): string {
    if (!this._reverseMap.has(key)) {
      throw new Error(`SymbolIds does not contain value: ${key}`);
    }
    return this._reverseMap.get(key) as string;
  }

  getPos(key: string): number {
    if (!this._pos.has(key)) {
      throw new Error(`SymbolIds does not contain key: ${key}`);
    }
    return this._pos.get(key) as number;
  }

  set(key: string, value: number, pos: number): void {
    this._map.set(key, value);
    this._pos.set(key, pos);
    this._reverseMap.set(value, key);
  }

  *[Symbol.iterator](): IterableIterator<[string, number]> {
    yield* [...this._map.entries()];
  }
}

File: packages/gbnf-experiment/data/20260909T133702Z_8305aa82/ported_implementation/gbnf/rules_builder/symbol_ids.py
from typing import Dict, Iterator, Tuple


class SymbolIds:
    """We don't need to delete, just preserve relationships."""

    def __init__(self):
        self._map: Dict[str, int] = {}
        self._pos: Dict[str, int] = {}
        self._reverse_map: Dict[int, str] = {}

    @property
    def size(self) -> int:
        return len(self._map)

    def keys(self):
        return list(self._map.keys())

    def has(self, key: str) -> bool:
        return key in self._map

    def get(self, key: str) -> int:
        if key not in self._map:
            raise Exception(f"SymbolIds does not contain key: {key}")
        return self._map[key]

    def reverse_get(self, key: int) -> str:
        if key not in self._reverse_map:
            raise Exception(f"SymbolIds does not contain value: {key}")
        return self._reverse_map[key]

    def get_pos(self, key: str) -> int:
        if key not in self._pos:
            raise Exception(f"SymbolIds does not contain key: {key}")
        return self._pos[key]

    def set(self, key: str, value: int, pos: int) -> None:
        self._map[key] = value
        self._pos[key] = pos
        self._reverse_map[value] = key

    def __iter__(self) -> Iterator[Tuple[str, int]]:
        return iter(list(self._map.items()))
